cpp/include/raft/spatial/knn/knn.cuh

```diff
@@ -168,6 +223,72 @@ inline void select_k(const value_t* in_keys,
   }
 }
 
+/**
+ * Select k smallest or largest key/values from each row in the input data.
+ *
+ * If you think of the input data `in_keys` as a row-major matrix with input_len columns and
+ * n_inputs rows, then this function selects k smallest/largest values in each row and fills
+ * in the row-major matrix `out_keys` of size (n_inputs, k).
+ *
+ * Note, depending on the selected algorithm, the values within rows of `out_keys` are not
+ * necessarily sorted. See the `SelectKAlgo` enumeration for more details.
+ *
+ * @tparam idx_t
+ *   the payload type (what is being selected together with the keys).
+ * @tparam value_t
+ *   the type of the keys (what is being compared).
+ *
+ * @param[in] handle the cuml handle to use
+ * @param[in] in_keys
+ *   contiguous device array of inputs of size (input_len * n_inputs);
+ *   these are compared and selected.
+ * @param[in] in_values
+ *   contiguous device array of inputs of size (input_len * n_inputs);
+ *   typically, these are indices of the corresponding in_keys.
+ *   You can pass `NULL` as an argument here; this would imply `in_values` is a homogeneous array
+ *   of indices from `0` to `input_len - 1` for every input and reduce the usage of memory
+ *   bandwidth.
+ * @param[out] out_keys
+ *   contiguous device array of outputs of size (k * n_inputs);
+ *   the k smallest/largest values from each row of the `in_keys`.
+ * @param[out] out_values
+ *   contiguous device array of outputs of size (k * n_inputs);
+ *   the payload selected together with `out_keys`.
+ * @param[in] k
+ *   the number of outputs to select in each input row.
+ * @param[in] select_min
+ *   whether to select k smallest (true) or largest (false) keys.
+ * @param[in] algo
+ *   the implementation of the algorithm
+ */
+template <typename idx_t = int, typename value_t = float>
+inline void select_k(const raft::handle_t& handle,
+                     raft::device_matrix_view<const value_t, idx_t, row_major> in_keys,
+                     raft::device_matrix_view<const idx_t, idx_t, row_major> in_values,
+                     raft::device_matrix_view<value_t, idx_t, row_major> out_keys,
+                     raft::device_matrix_view<idx_t, idx_t, row_major> out_values,
+                     int k,
+                     bool select_min  = true,
```

Comment:
If `select_min` and `algo` are algorithm parameters that have reasonable defaults, then would you consider gathering them into a struct, in order to avoid potential mix-ups with the order of `k` and `select_min`?  Alternately (or additionally), would you consider making `select_min` an `enum class` instead of a `bool`, so it can't be mixed up with `k`?

Reply:
Again proposing separating out into TWO APIs, `select_k_max` and `select_k_min`

Reply:
I think I might actually remove the `select_k` for now because we're in the process of pulling the individual algorithms out of `detail` and moving them to `raft::matrix`. When that happens, I don't think `select_k` belongs in `raft::spatial`.